pkg/service/grantsvc/granter.go

```diff
@@ -137,6 +137,47 @@ func (g *Granter) RevokeGrant(ctx context.Context, opts RevokeGrantOpts) (*acces
 	return nil, errors.New("unhandled response code")
 }
 
+// validate grant runs all the checks that will need to occur when creating a real grant to validate its success
+func (g *Granter) ValidateGrant(ctx context.Context, opts CreateGrantOpts) error {
+	q := &storage.GetUser{
+		ID: opts.Request.RequestedBy,
+	}
+	_, err := g.DB.Query(ctx, q)
+	if err != nil {
+		return err
+	}
+	start, end := opts.Request.GetInterval(access.WithNow(g.Clock.Now()))
+
+	req := ahTypes.ValidateRequestToProviderJSONRequestBody{
+		Id:       opts.Request.ID,
+		Provider: opts.AccessRule.Target.ProviderID,
+		With: ahTypes.CreateGrant_With{
+			AdditionalProperties: make(map[string]string),
+		},
+		Subject: openapi_types.Email(q.Result.Email),
+		Start:   iso8601.New(start),
+		End:     iso8601.New(end),
+	}
+	for k, v := range opts.AccessRule.Target.With {
+		req.With.AdditionalProperties[k] = v
+	}
+	for k, v := range opts.Request.SelectedWith {
+		req.With.AdditionalProperties[k] = v.Value
+	}
+
+	res, err := g.AHClient.ValidateRequestToProviderWithResponse(ctx, opts.AccessRule.Target.ProviderID, req)
```

Comment:
It looks like the response here from validateRequest is discarded.

This means that if there are any validation errors they will be ignored!

I think you will need to expand this to check the contents of the validation response body and report the errors back to the client.